pennylane/templates/embeddings/amplitude.py

```diff
@@ -15,27 +15,131 @@
 Contains the ``AmplitudeEmbedding`` template.
 """
 # pylint: disable-msg=too-many-branches,too-many-arguments,protected-access
-import math
+import warnings
 import numpy as np
+
+import pennylane as qml
 from pennylane.templates.decorator import template
 from pennylane.ops import QubitStateVector
-from pennylane.templates.utils import (
-    check_shape,
-    check_type,
-    get_shape,
-)
 from pennylane.variable import Variable
 from pennylane.wires import Wires
+from pennylane.templates.utils import check_shape, get_shape, check_type
 
 # tolerance for normalization
 TOLERANCE = 1e-10
 
 
+def _preprocess(features, wires, pad_with, normalize):
+    """Validate and pre-process inputs."""
+
+    if qml.tape_mode_active():
+
+        shape = qml.math.shape(features)
+
+        # check shape
+        if len(shape) != 1:
+            raise ValueError(f"Features must be a one-dimensional vector; got shape {shape}.")
+
+        n_features = shape[0]
+        if pad_with is None and n_features != 2 ** len(wires):
+            raise ValueError(
+                f"Features must be of length {2 ** len(wires)}; got length {n_features}. "
+                f"Use the 'pad' argument for automated padding."
+            )
+
+        if pad_with is not None and n_features > 2 ** len(wires):
+            raise ValueError(
+                f"Features must be of length {2 ** len(wires)} or "
+                f"smaller to be padded; got length {n_features}."
+            )
+
+        # pad
+        if pad_with is not None and n_features < 2 ** len(wires):
+            padding = [pad_with] * (2 ** len(wires) - n_features)
+            features = qml.math.concatenate([features, padding], axis=0)
+
+        # normalize
+        norm = qml.math.sum(qml.math.abs(features) ** 2)
+
+        if not qml.math.allclose(norm, 1.0, atol=TOLERANCE):
+            if normalize or pad_with:
+                features = features / np.sqrt(norm)
```

Comment:
As discussed, JAX abstract types will break on conditional branching based on tensor values.

Reply:
We need a general strategy how to check numerical properties of inputs then, for example that inputs of BasisEmbedding are 0/1, that inputs to AmplitudeEmbedding are normalised etc...

Reply:
* For basis embedding, since it is not differentiable, what if we restricted it to only NumPy arrays? Would the computational graph just treat is as a constant?

* For amplitude embedding, one solution is to simply always normalize (e.g., just get rid of the if statement).